app/controllers/concerns/current_organizationable.rb

```diff
@@ -0,0 +1,55 @@
+module CurrentOrganizationable
+  extend ActiveSupport::Concern
+
+  def current_organization
+    return unless current_user
+
+    @current_organization ||= set_current_organization
+  end
+
+  private
+
+  def set_current_organization
+    session[:organization_id] = organization&.id
+    Current.organization = organization
+
+    organization
+  end
+
+  def organization
+    @organization ||= if session[:organization_id].nil?
+      set_default_organization
+    else
+      find_organization
+    end
+  end
+
+  def find_organization
+    id = organization_id if valid_user_organization?(organization_id)
+    Organization&.find(id)
+  end
+
+  def organization_id
+    return params["current-organization"] if params["current-organization"]
+
+    if controller_name.inquiry.organizations? && params[:action] != "index"
+      params[:id]
+    else
+      session[:organization_id] || params[:organization_id]
+    end
+  end
+
+  def set_default_organization
+    if current_user.organization_users.nil?
+      current_user&.organization
+    else
+      current_user.default_organization
+    end
+  end
+
+  def valid_user_organization?(org_id)
+    return true if authorized_user.can_admin_system?
+
+    current_user.organizations.map(&:id).include?(org_id.to_i)
+  end
+end
```

Comment:
Much of the logic in this concern seems like it could be condensed to improve readability with a more linear/procedural approach. Consider some of the side by side comparison critiques of the Sandi Metz OOP style of lots of little methods vs a more linear approach using local variables. I expect that if you reworked this module, you'd find that you can reduce the code by half or more while making it more understandable. Anyway, it's not a hill I'd die on right now but will simply encourage you to try the thought experiment and draw your own conclusions.

Reply:
I added some changes...not sure if I made it better or worse. There were a lot of edge cases I encountered which is why this logic got as large as it did. I have actually refactored it several times at this point. Happy to take suggestions if you are still unhappy.